zero_functional.nim

```diff
@@ -435,21 +446,31 @@ proc iterHandler(args: NimNode): NimNode {.compileTime.} =
   var code = result[^1]
   let initials = nnkStmtList.newTree()
   result[^1].add(initials)
+  var forceSeq = false
 
   if args.len > 0 and args[^1].len > 0:
     let lastCall = $args[^1][0]
     if lastCall in SEQUENCE_HANDLERS:
+      forceSeq = lastCall.endsWith("Seq")
       let resultIdent = newIdentNode("result")
-      let resultType = compileTimeTypeInfer(args)
-      let zero = quote:
-        `resultIdent` = `resultType`
+      let zero = 
+        if not forceSeq:
+          let resultType = compileTimeTypeInfer(args)
+          quote:
+            `resultIdent` = `resultType`
+        else:
+          let emptyIdent = newIdentNode(emptyIdentifier)
```

Comment:
had to use the emptyIdent and building the sequence from first element again, because `compileTimeTypeInfer` would not really work on arbitrary iterators

Reply:
Can you give an example of an iterator which has problems with compileTypeInfer

Reply:
OK - this is a somewhat constructed example, but if I define my own iterator - like this:

```nim
type Pack = ref object
  rows: seq[int]
proc len(pack: Pack) : int = 
  pack.rows.len()
proc `[]`(pack: Pack, idx: int) : int  = 
  pack.rows[idx]
iterator items(a: Pack): int =
  ## iterates over each item of `a`.
  var i = 0
  let L = len(a)
  while i < L:
    yield a[i]
    inc(i)
    assert(len(a) == L, "seq modified while iterating over it")
```
then this code will compile fine (with the emptyIdent added again):
```nim
    let p = Pack(rows: @[0,1,2,3])
    echo(p --> filterSeq(it != 0))    
```
This one will not:
```nim 
    let p = Pack(rows: @[0,1,2,3])
    echo(p --> filter(it != 0))    
```